backend_new/src/services/application-flagged-set.service.ts

```diff
@@ -0,0 +1,951 @@
+    builds the ruleKey field given an application, the rule, and the listingId
+  */
+  buildRuleKey(
+    application: Application,
+    rule: RuleEnum,
+    listingId: string,
+  ): string {
+    if (rule == RuleEnum.email) {
+      return `${listingId}-email-${application.applicant.emailAddress}`;
+    } else {
+      return (
+        `${listingId}-nameAndDOB-${application.applicant.firstName}-${application.applicant.lastName}-${application.applicant.birthMonth}-` +
+        `${application.applicant.birthDay}-${application.applicant.birthYear}`
+      );
+    }
+  }
+
+  /**
+    gathers a set of applications that match the passed in application based on the rule
+  */
+  async checkForMatchesAgainstRule(
+    application: Application,
+    rule: RuleEnum,
+    listingId: string,
+  ): Promise<Application[]> {
+    if (rule === RuleEnum.email) {
+      return await this.checkAgainstEmail(application, listingId);
+    } else if (rule === RuleEnum.nameAndDOB) {
+      return await this.checkAgainstNameAndDOB(application, listingId);
+    }
+  }
+
+  /**
+    gets a list of applications that matches based on email
+  */
+  async checkAgainstEmail(
+    application: Application,
+    listingId: string,
+  ): Promise<Application[]> {
+    if (!application?.applicant?.emailAddress) {
+      return [];
+    }
+
+    const apps = await this.prisma.applications.findMany({
+      select: {
+        id: true,
+      },
+      where: {
+        id: {
+          not: application.id,
+        },
+        status: ApplicationStatusEnum.submitted,
+        listingId: listingId,
+        applicant: {
+          emailAddress: application.applicant.emailAddress,
+        },
+      },
+    });
+
+    return mapTo(Application, apps);
+  }
+
+  /**
+    gets a list of applications that matches based on name and dob (spread across applicant + householdmember)
+  */
+  async checkAgainstNameAndDOB(
+    application: Application,
+    listingId: string,
+  ): Promise<Application[]> {
+    const firstNames = [
+      application.applicant.firstName,
+      ...(application.householdMember
+        ? application.householdMember.map((member) => member.firstName)
+        : []),
+    ];
+    const lastNames = [
+      application.applicant.lastName,
+      ...(application.householdMember
+        ? application.householdMember.map((member) => member.lastName)
+        : []),
+    ];
+    const birthMonths = [
+      application.applicant.birthMonth,
+      ...(application.householdMember
+        ? application.householdMember.map((member) => member.birthMonth)
+        : []),
+    ];
+    const birthDays = [
+      application.applicant.birthDay,
+      ...(application.householdMember
+        ? application.householdMember.map((member) => member.birthDay)
+        : []),
+    ];
+    const birthYears = [
+      application.applicant.birthYear,
+      ...(application.householdMember
+        ? application.householdMember.map((member) => member.birthYear)
+        : []),
```

Comment:
suggestion (non-blocking): you could create a function to handle each of these that accepted a "key" parameter like "lastName" and returned the array.

Reply:
nice call out! updated